### PR TITLE
[Core] Extend NormalCalculationUtils. Fix #7052

### LINKED ISSUE #7052: Normal computation on condition using CalculateNormal3D for 4 noded (quad)
Hi All, I was wondering if the functionality of the CalculateNormal3D in NormalCalculationUtils could be extended to linear quads. Currently, it considers only 3 nodes in the normal computation.

Any help/thought much appreciated, thank you.

-----------------------------------------------
@ddiezrod 
@pooyan-dadvand

### LINKED ISSUE #7178: Adding test for calculation of normals
to avoid the problems that were fixed related to the synchronization of the nodal flags

The test should be i c++ imo and check for the same results in openmp and MPI

### PR DESCRIPTION
**Pull request by loumalouomega**
Fixes #7052
closes #7178 

This adds utilities to compute the normals in any type of geometry, not just triangles and lines. It checks if the simplest geometries are available and uses the simplest, otherwise uses generic algorithm

It adds a quad test

I will clean up MortarUtilities in a future PR

### CODE REVIEW

**Comment by loumalouomega** [comment]
BTW, we should think about adding MPI tests @philbucher

**Comment by philbucher** [comment]
> BTW, we should think about adding MPI tests @philbucher

I completely agree!

I will think about it while reviewing (on vacation atm)

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]: Would it be possible to refactor this normal calculation, one calculates normals on conditions and stores the normal in condition data value container, then another method spreads it among nodes from conditional normals?

This refactoring will be very useful if someone wants condition normal for Neumann conditions, and if he doesn't want to add NORMAL to solution step variables.

Comment (sunethwarna) [reply kratos/utilities/normal_calculation_utils.cpp]: The example is, I am using CalculateOnSimplex for my whole model part since I need condition normals for all the conditions in specific RANS formulations. After that, I am spreading nodal normals to which SLIP flag is true, to have SLIP condition for flow solver.

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
            const double coefficient = 1.0/static_cast<double>(r_geometry.PointsNumber());
```

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]: What if we do OMP parallelism using node locks in here?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]: Again the OMP parallism, would there be a problem if we use it with node locks?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
        const double coeff = 1.0/pGeometry.size();
```

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    itCond->SetValue(NORMAL, rAn);
```
This will also work ryt?, it will be helpfull in OMP parallelism. :)

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    itCond->SetValue(NORMAL, rAn);
```

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp:119]: Would it be good to have KRATOS_TRY/KRATOS_CATCH here since you are having an error thrown so the location of the error can be determined easily with full debug?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (sunethwarna) [inline kratos/utilities/normal_calculation_utils.cpp:233]: May be KRATOS_TRY/KRATOS_CATCH?

**Review by sunethwarna — COMMENTED** [review]
I have a general question,

Would it be possible to refactor Normal calculation,
1. Calculates normals on conditions in one method and stores in Condition(Geometry)DataValueContainer
2. A method to spread condition normals to nodal normals based on SLIP flag.
So in the existing method calls can utilize both of above mentioned methods to have exiting functionality, and if someone wants to calculate NORMALS on conditions without harming nodal NORMALS so method 1 could be used.

Other thing is, OMP parallelism is missing in almost all of the loops for somereason, may be we are using copy paste of old code?

I only have some minor comments other than that. :). Thanks @loumalouomega  :)

**Comment by sunethwarna** [comment]
BTW, windows build is failing in CI

**Comment by loumalouomega** [comment]
> 
> 
> BTW, windows build is failing in CI

I am on it

File: kratos/utilities/normal_calculation_utils.cpp
Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.cpp]: I will not, I am computing normals in SLIP flag nodes, please do as I do an create a submodelpart. I don't want to  modify current approach

Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.cpp]: > Would it be possible to refactor this normal calculation, one calculates normals on conditions and stores the normal in condition data value container, then another method spreads it among nodes from conditional normals?

Let's see

**Comment by philbucher** [comment]
check #6669 to see how the template stuffs can be done to satisfy all compilers

**Comment by philbucher** [comment]
maybe looking at [this file](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MappingApplication/custom_utilities/mapper_factory.cpp) helps too

**Comment by loumalouomega** [comment]
Looks like I have fixed it

**Comment by philbucher** [comment]
> Looks like I have fixed it @philbucher

Hm I am not sure if using ifdef is really a fix 😅

**Comment by loumalouomega** [comment]
> > Looks like I have fixed it @philbucher
> 
> Hm I am not sure if using ifdef is really a fix sweat_smile

imagen

File: kratos/utilities/normal_calculation_utils.cpp
Comment (mohittyagi-k) [inline kratos/utilities/normal_calculation_utils.cpp]: This feels like too much to grasp from one line.  Also, this seems to be the heart of the PR, so probably it would be good to have it in the following form:
```
    if (EnforceGenericAlgorithm)
    {
        use_simplex = false;
    }
    else if (dimension == 2)
    {
        use_simplex = (geomtype == line2d2);
    }
    else
    {
        use_simplex = (geomtype == triangle3d3);
    }
```

File: kratos/utilities/normal_calculation_utils.cpp
Comment (mohittyagi-k) [inline kratos/utilities/normal_calculation_utils.cpp:179]: what if CalculateOnSimplex is called on quads?

Comment (ddiezrod) [reply kratos/utilities/normal_calculation_utils.cpp:179]: Only triangles and tets are simplex. This specific method is maintained for performance reasons.

Comment (mohittyagi-k) [reply kratos/utilities/normal_calculation_utils.cpp:179]: yes, so this could be an error message. Just to avoid the mistake. In any case, this is fine with me.

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee ?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.cpp]: well to me this is not sth that should go to master. God know what consequences this has in the different OSs
I pointed to the other files that do the same, can you take a look please?

**Review by philbucher — COMMENTED** [review]
why adding another large mdpa? Cannot you reuse an existing one?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    auto& r_nodes_array = rModelPart.GetCommunicator().LocalMesh().Nodes();
```

otherwise it won't work in MPI

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp:119]: I agree

File: kratos/utilities/normal_calculation_utils.h
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.h]: what is a generic algorithm?
Please use something more meaningful

File: kratos/utilities/normal_calculation_utils.h
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.h]: if this has to be 2 or 3 then the default cannot be 0!
```suggestion
        const std::size_t Dimension
```

**Comment by loumalouomega** [comment]
> why adding another large mdpa? Cannot you reuse an existing one?

Current one is triangles, proposed is quadrilaterals

File: kratos/utilities/normal_calculation_utils.cpp
Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.cpp]: OK

File: kratos/utilities/normal_calculation_utils.h
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.h]: since this is a big refactor can you please directly use a ´const Condition&` here?

Or if this is still needed for backward compatibility make a second function and deprecate this one

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp:233]: agree

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.cpp]: why is this different for elements & conditions?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.cpp]: same

File: kratos/tests/test_normal_utils.py
Comment (philbucher) [inline kratos/tests/test_normal_utils.py:83]: this test must also be run in MPI and the results have to be the same, given that we recently had problems

@roigcarlo is this already #7178?

once you are done with the changes I can port it such that it works/runs both with and wthout mpi

**Review by philbucher — COMMENTED** [review]
> > why adding another large mdpa? Cannot you reuse an existing one?
> 
> Current one is triangles, proposed is quadrilaterals

but why so large?

**Comment by loumalouomega** [comment]
> > > why adding another large mdpa? Cannot you reuse an existing one?
> > 
> > 
> > Current one is triangles, proposed is quadrilaterals
> 
> but why so large?

It is quite coarse, and I removed submodelparts

File: kratos/utilities/normal_calculation_utils.cpp
Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.cpp]: Look at the code ad you will see why

File: kratos/utilities/normal_calculation_utils.h
Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.h]: 0 means that it must be retrieved from DOMAIN_SIZE

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp]: I did and I don't understand it
All we are interested is Geometries, doen't matter if its an element or a condition

**Comment by philbucher** [comment]
> > > > why adding another large mdpa? Cannot you reuse an existing one?
> > > 
> > > 
> > > Current one is triangles, proposed is quadrilaterals
> > 
> > 
> > but why so large?
> 
> It is quite coarse, and I removed submodelparts

but why do you need almost 1000 conditions for a unittest for normals calculation? :sweat_smile: 
Shouldn't 10 or so be enough to test all sort of different combinations?

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp]: how about the following (using names just to show my idea):
Interface:
- `CalculateElementNormals`
- `CalculateConditionNormals`

in the cpp: `template<TEntityType> CalculateCondition(Entities)`
=> this method is called from `CalculateElementNormals` and `CalculateConditionNormals`

**Comment by mohittyagi-k** [comment]
@loumalouomega  eagerly awaiting some buzz on this PR? 😄

File: kratos/tests/test_normal_utils.py
Comment (roigcarlo) [reply kratos/tests/test_normal_utils.py:83]: For MPI we will have to test particular corner cases, but agree with you, this should also run in MPI and give the same results. 

I am taking a look at the new implementation in case I see something odd with mpi.

Comment (ddiezrod) [reply kratos/tests/test_normal_utils.py:83]: The tests we do in @KratosMultiphysics/altair follow this structure so they can be run both in mpi and openmp.

https://github.com/KratosMultiphysics/Kratos/blob/48725f2120182b2c19e8aac49665d0d82f10118e/applications/FluidDynamicsApplication/tests/mass_conservation_check_process_test.py

What do you think? @philbucher

Comment (philbucher) [reply kratos/tests/test_normal_utils.py:83]: didn't I suggest to do it like this in the first place :D

In any case I wanted for a while to unify sth like this in the core => separate PR (looking @sunethwarna for someone to help me with this :smiley:)

Comment (philbucher) [reply kratos/tests/test_normal_utils.py:83]: we will get back to you asap, I think this PR could really benefit from it!

Comment (philbucher) [reply kratos/tests/test_normal_utils.py:83]: @sunethwarna and I discussed, we will get back to you by the end of the week with sth that can run the test both in serial and in MPI

**Comment by sunethwarna** [comment]
Hi I added the `KratosCoreTestCase` class to enable these tests to run in MPI as well as discussed with @philbucher. If you like, I can commit the changes I did in `test_normal_utils.py` file so same tests can be run in MPI as well. (I used this PR as the experimental setup to create `KratosCoreTestCase`)

**Comment by sunethwarna** [comment]
@loumalouomega Hi, If you like, I can push the tests so same  tests can be run in MPI as well ? I also need this branch to go through.

File: kratos/utilities/normal_calculation_utils.h
Comment (RiccardoRossi) [reply kratos/utilities/normal_calculation_utils.h]: i agree with Vicrntr on this. initializing from an impodsible number obliges to read

Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.h]: Cannot be const as it sets the NORMAL

File: kratos/utilities/normal_calculation_utils.cpp
Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.cpp]: This avoids two declarations in the header

**Comment by loumalouomega** [comment]
Test in MPI added (@philbucher I don't know how to add MPI tests in the core, as in the applications?)

File: kratos/tests/test_normal_utils_mpi.py
Comment (philbucher) [inline kratos/tests/test_normal_utils_mpi.py]: @loumalouomega you misunderstood a bit how the testing in MPI is done

The point is that THE SAME test has to be run in order to make sure the results are consistent in shared and distributed memory => the only difference is the reading of the ModelPart

Take a look at how @sunethwarna has done it in #7345

I will take a closer look so that this can go in when I am back from vacations in two weeks

Comment (loumalouomega) [reply kratos/tests/test_normal_utils_mpi.py]: Well,  @sunethwarna is going to fix it (I am still on vacation). The important thing is that the utilities work in MPI, it is just a matter of adapting the test

Comment (sunethwarna) [reply kratos/tests/test_normal_utils_mpi.py]: done :)

Comment (loumalouomega) [reply kratos/tests/test_normal_utils_mpi.py]: Hurray

**Comment by loumalouomega** [comment]
@philbucher This can proceed

File: kratos/mpi/tests/test_KratosMPICore.py
Comment (philbucher) [inline kratos/mpi/tests/test_KratosMPICore.py:23]: nice!

File: kratos/utilities/normal_calculation_utils.h
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.h]: can you then please add a second function that calls this one with the dimension extracted from the DOMAIN_SIZE, such that the usage is crystal clear

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp:233]: ping

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp]: ping

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [reply kratos/utilities/normal_calculation_utils.cpp]: ping

**Review by philbucher — COMMENTED** [review]
the testing is very good now but there is still some major things in the implementation, esp the duplicated code for using Element/Condition => this should be templated such that that the logic/the algorithm is only implemented once, as I pointed out

File: kratos/utilities/normal_calculation_utils.cpp
Comment (philbucher) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
        for(ConditionsArrayType::iterator it =  rConditions.begin(); it !=rConditions.end(); i++t) {
```

Comment (sunethwarna) [reply kratos/utilities/normal_calculation_utils.cpp]: done :)

**Review by philbucher — COMMENTED** [review]
looks very good now, thx

I will approve on Monday so that others have time to also take a look
@roigcarlo @ddiezrod I think you guys should have a look too just to be sure

@sunethwarna I think you are still missing some KRATOS_TRY/CATCH in some functions

File: kratos/tests/test_redistance.py
Comment (ddiezrod) [inline kratos/tests/test_redistance.py]: Why is this here? Was the test failing?

Comment (sunethwarna) [reply kratos/tests/test_redistance.py]: yeah, I had to add Conditions Blocks to mdpas (coarse_sphere.mdpa). Afterwards it started to fail. I checked, existing DISTANCE calculator doesn't use NORMAL variable, or any normal_calculation_utils methods. So this should be because of change in mdpas.

As I see, VariationalDistanceCalculation Process is run only with 2 max iterations, which is not enough for convergent solution. So I increased number of max iterations untill convegent solution is given. Once that solution is obtained, I used that value. Do you also have some tests which uses VariationalDistanceCalculationProcess? If so could you please also check.

Comment (sunethwarna) [reply kratos/tests/test_redistance.py]: I checked with current master as well, In there, if you increase `max_iterations`, then It will again fail. But it will pass with increased amount of iterations with the value I have provided in this PR.

https://github.com/KratosMultiphysics/Kratos/blob/1d5ddbd3b218c8e7e3cb49566ede6d4898fd3453/kratos/tests/test_redistance.py#L40-L51

Comment (ddiezrod) [reply kratos/tests/test_redistance.py]: I agree that 2 are too few iterations but I wouldn't change it here. If test fails with max_iterations=2, something is wrong somewhere.

Comment (sunethwarna) [reply kratos/tests/test_redistance.py]: The difference is there because, now we have conditions block in mdpa, it marks nodes belonging to conditions with `BOUNDARY` = `true`. And in `DistanceCalculationElementSimplex`, it adds additional `RHS` contribution when `BOUNDARY` of a node is true as in following line.

https://github.com/KratosMultiphysics/Kratos/blob/1d5ddbd3b218c8e7e3cb49566ede6d4898fd3453/kratos/elements/distance_calculation_element_simplex.h#L260-L264

This is the reason for change in reference values. This difference in reference values can be eliminated by having a new PR to increase `max_iterations` and then getting the reference values from that to pass this test since at the convergence those values are the same.

I don't think there is something wrong happening in here, its only the approach for convergence is altered when conditions block is added to mdpa. For further checking, I have created PR #7389. Once that is merged, then we can revert `test_redistance.py` changes in this PR.

Comment (sunethwarna) [reply kratos/tests/test_redistance.py]: After discussing with @philbucher , I reverted changes done to existing mdpas and `test_redistance.py` and introduced `coarse_spehere_with_conditions.mdpa` to be used in `test_normal_utils.py`.

Comment (ddiezrod) [reply kratos/tests/test_redistance.py]: that makes sense to me. Thanks!

Comment (loumalouomega) [reply kratos/tests/test_redistance.py]: What I do is using SkinDetectionProcess instead of creating a new mdpa

Comment (loumalouomega) [reply kratos/tests/test_redistance.py]: > After discussing with @philbucher , I reverted changes done to existing mdpas and `test_redistance.py` and introduced `coarse_spehere_with_conditions.mdpa` to be used in `test_normal_utils.py`.

I was using SkinDetectionProcess, why did you change it?

Comment (philbucher) [reply kratos/tests/test_redistance.py]: Although usually I would say that it is better to re-use things I wold say that in this case it makes sense to have another file that is self contained and we know that it works

Using the skin-detection process adds another layer of complexity to an already existing complex case
Plus then we need to write and re-read the mdpa file that the process generates which is also not an optimal solution

maybe once the skin detection also works in MPI we could come back to this discussion and think if we want to use it, but for the moment I would not

Comment (loumalouomega) [reply kratos/tests/test_redistance.py]: Yes, it is something want to do in the sort term

**Comment by sunethwarna** [comment]
Can we merge this if everything is ok? @loumalouomega @philbucher @mohittyagi-k @ddiezrod

**Review by philbucher — APPROVED** [review]
to me this is ready, adds proper tests and has been open for a while

Thx @loumalouomega @sunethwarna

**Pull request merged**
